discovery/eureka/eureka.go

```diff
@@ -0,0 +1,216 @@
+)
+
+const (
+	presentValue = model.LabelValue("true")
+
+	// metaLabelPrefix is the meta prefix used for all meta labels.
+	// in this discovery.
+	metaLabelPrefix = model.MetaLabelPrefix + "eureka_"
+
+	appNameLabel                     = metaLabelPrefix + "app_name"
+	appInstanceHostNameLabel         = metaLabelPrefix + "app_instance_hostname"
+	appInstanceIPAddrLabel           = metaLabelPrefix + "app_instance_ipaddr"
+	appInstanceStatusLabel           = metaLabelPrefix + "app_instance_status"
+	appInstanceIDLabel               = metaLabelPrefix + "app_instance_id"
+	appInstanceMetadataPrefix        = metaLabelPrefix + "app_instance_metadata_"
+	appInstanceMetadataPresentPrefix = metaLabelPrefix + "app_instance_metadatapresent_"
+)
+
+// DefaultSDConfig is the default Eureka SD configuration.
+var DefaultSDConfig = SDConfig{
+	RefreshInterval: model.Duration(30 * time.Second),
+}
+
+// SDConfig is the configuration for applications running on Eureka.
+type SDConfig struct {
+	Servers          []string                     `yaml:"servers,omitempty"`
+	RefreshInterval  model.Duration               `yaml:"refresh_interval,omitempty"`
+	HTTPClientConfig config_util.HTTPClientConfig `yaml:",inline"`
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+	if len(c.Servers) == 0 {
+		return errors.New("eureka_sd: must contain at least one Eureka server")
+	}
+
+	return c.HTTPClientConfig.Validate()
+}
+
+type applicationsClient func(ctx context.Context, servers []string, client *http.Client) (*Applications, error)
+
+// Discovery provides service discovery based on a Eureka instance.
+type Discovery struct {
+	*refresh.Discovery
+	client      *http.Client
+	servers     []string
+	lastRefresh map[string]*targetgroup.Group
+	appsClient  applicationsClient
+}
+
+// New creates a new Eureka discovery for the given role.
+func NewDiscovery(conf SDConfig, logger log.Logger) (*Discovery, error) {
+	rt, err := config_util.NewRoundTripperFromConfig(conf.HTTPClientConfig, "marathon_sd", false)
+	if err != nil {
+		return nil, err
+	}
+
+	d := &Discovery{
+		client:     &http.Client{Transport: rt},
+		servers:    conf.Servers,
+		appsClient: fetchApps,
+	}
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"eureka",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d, nil
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	targetMap, err := d.fetchTargetGroups(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	all := make([]*targetgroup.Group, 0, len(targetMap))
+	for _, tg := range targetMap {
+		all = append(all, tg)
+	}
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+
+	// Remove services which did disappear.
+	for source := range d.lastRefresh {
```

Comment:
Why is that needed?

Reply:
If the application restarts, it may get out of Discovery.
To avoid immediate removal from the monitoring target if removed from the Discovery.

Reply:
This is not a correct behavior from the Prometheus perspective. Can you remove this, + lastRefresh?

Reply:
ok. I will remove it